ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -598,45 +600,67 @@ def get_physical_pages(self, for_fileIds=None, for_pageIds=None, return_divs=Fal
         if for_pageIds is not None:
             ret = []
-            pageId_patterns = []
+            page_attr_patterns = []
             for pageId_token in re.split(r',', for_pageIds):
                 if pageId_token.startswith(REGEX_PREFIX):
-                    pageId_patterns.append(re.compile(pageId_token[REGEX_PREFIX_LEN:]))
+                    page_attr_patterns.append(re.compile(pageId_token[REGEX_PREFIX_LEN:]))
                 elif '..' in pageId_token:
-                    pageId_patterns += generate_range(*pageId_token.split('..', 1))
+                    page_attr_patterns += generate_range(*pageId_token.split('..', 1))
                 else:
-                    pageId_patterns += [pageId_token]
-            if self._cache_flag:
-                for page_id in self._page_cache.keys():
-                    if page_id in pageId_patterns or \
-                            any([isinstance(p, typing.Pattern) and p.fullmatch(page_id) for p in pageId_patterns]):
-                        if return_divs:
-                            ret.append(self._page_cache[page_id])
-                        else:
-                            ret.append(page_id)
-            else:
-                for page in self._tree.getroot().xpath(
-                        'mets:structMap[@TYPE="PHYSICAL"]/mets:div[@TYPE="physSequence"]/mets:div[@TYPE="page"]',
-                        namespaces=NS):
-                    page_id = page.get('ID')
-                    if page_id in pageId_patterns or \
-                            any([isinstance(p, typing.Pattern) and p.fullmatch(page_id) for p in pageId_patterns]):
-                        if return_divs:
-                            ret.append(page)
-                        else:
-                            ret.append(page_id)
+                    page_attr_patterns += [pageId_token]
+            if page_attr_patterns:
+                if self._cache_flag:
+                    # determine attr to look for before iterating
+                    try:
+                        attr = next(a for a in METS_PAGE_DIV_ATTRIBUTE if (
+                                    any(p in self._page_cache[a] for p in page_attr_patterns) or \
+                                    any([isinstance(p, typing.Pattern) and p.fullmatch(attr_val) \
+                                        for p in page_attr_patterns \
+                                        for attr_val in self._page_cache[a]]
+                                    )))
+                        for attr_val in self._page_cache[attr].keys():
+                            if attr_val in page_attr_patterns or \
+                                    any([isinstance(p, typing.Pattern) and p.fullmatch(attr_val) for p in page_attr_patterns]):
+                                if return_divs:
+                                    ret.append(self._page_cache[attr][attr_val])
+                                else:
+                                    ret.append(attr_val)
```

Comment:
It currently drops the range in both cases, did not anticipate this, added a test for it and will add another loop to prevent this behavior. Thanks for spotting this.